### PR TITLE
Add windows service functionality #344

### PR DESCRIPTION
**Pull request by MolotovCherry**
This is an initial implementation of a Windows service for pueued. #344 

~~This is a draft and I would like to have some testing/feedback on it before it gets merged.~~

How to use:
The following subcommands now exist:
* `pueued service install`
* `pueued service uninstall`
* `pueued service start`
* `pueued service stop`
* `pueued service run` (internal only)

In an admin terminal:
* `pueued service install`
* Go to your services manager and start the service, or run `pueued service start`. It's set to autostart, so it will start up with the system, though that can obviously be adjusted to manual if the user wants. After the initial install, _the service is not started automatically_, so you must start it yourself the first time.
* To uninstall `pueued service uninstall`

image

This should also respect it when users log out and other users log in.

The service process itself runs as SYSTEM. It spawns a new `pueued` daemon process as the currently logged in user.

What I want tested:
- Just the general functionality, but specifically I want to know how it works when:
- [x] - user logs out
- [x] - another user logs in
- [x] - on bootup/restart

Todo / possible bug:
- [x] What happens if the daemon process fails on its own?
- [x] Adding tasks then trying to follow / log them / status, gives these weird "cannot deserialize" errors. Possibly caused by not enough perms? Should they be inherited then? - Edit: not my service's fault.
- [x] unpark inside stop after process killed so it doesn't get stuck on parked without ever being able to stop service
- [x] logoff/logon stops service
- [x] Fix daemon `-d` to actually detach in windows using process creation flags
- [x] Fix daemon in `-d` launching visible shell windows

Possible additional features:
- [ ] Make `-d` mode check if windows service exists, if so, start and run the service instead of the daemon, otherwise, fallback to forking daemon mode.

## Checklist

Please make sure the PR adheres to this project's standards:

- [x] I included a new entry to the `CHANGELOG.md`.
- [x] I checked `cargo clippy` and `cargo fmt`. The CI will fail otherwise anyway.
- [x] (If applicable) I added tests for this feature or adjusted existing tests.
- [x] (If applicable) I checked if anything in the wiki needs to be changed.*
* Following pages need to be changed:
  * https://github.com/Nukesor/pueue/wiki/Common-Pitfalls-and-Debugging#windows-quirks (this service fixes the quirk, though it should be noted the quirk is still there in `-d` mode, but using the service will properly fix it)
  * https://github.com/Nukesor/pueue/wiki/Get-started#start-the-daemon (add a section for windows and the service)

### CODE REVIEW

**Comment by MolotovCherry** [comment]
@Nukesor I am getting a lot of confusing errors, like the following. Have any clue what's up?
```
~
~\Downloads\vaultwarden
❯ pueue status
Error: Couldn't deserialize message:
invalid type: map, expected variant identifier

❯ pueue log
Failed to deserialize message: missing field `tasks`
```

**Comment by MolotovCherry** [comment]
Ok, I think it's good to go for review now.

**Comment by MolotovCherry** [comment]
I've also fixed `-d` on Windows. This command properly detaches from the parent process (with no console for parent or children), so closing the window is fine.

Since `-d` is fixed separately, technically the service itself probably isn't needed. A person _could_ just set up a startup or taskscheduler job to make it run on startup instead.

I can nuke the service out of this PR if you want.

**Comment by Nukesor** [comment]
Keep it in, I think that's the main reason why many people wanted to have the service in the first place :)

Code looks clean on well-documented on the first glance! Much appreciated. The windows portion of Pueue needed a bit of love since quite some time.

File: pueue/src/daemon/service.rs [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+
+#[derive(Clone)]
+struct Config {
+    config_path: Option<PathBuf>,
+    profile: Option<String>,
+}
+
+const SERVICE_NAME: &str = "pueued";
+const SERVICE_TYPE: ServiceType = ServiceType::OWN_PROCESS;
+static CONFIG: OnceCell<Config> = OnceCell::new();
```
Comment (simonsan) [inline pueue/src/daemon/service.rs:61]: Not sure once_cell is needed anymore as a dependency?

https://www.reddit.com/r/rust/comments/126qaai/after_years_of_work_and_discussion_once_cell_has/

Comment (MolotovCherry) [reply pueue/src/daemon/service.rs:61]: You'll need to [raise your msrv to 1.80](https://github.com/Nukesor/pueue/blob/1c5f4aa06d1ffa9081d3eea40b77d0e5d0d137cd/Cargo.toml#L14) before we can use `LazyLock`

Comment (simonsan) [reply pueue/src/daemon/service.rs:61]: @Nukesor Do you think it would be a problem to do that, I don't see an MSRV policy in the Readme, so not sure if there is a policy on this?

Comment (Nukesor) [reply pueue/src/daemon/service.rs:61]: Not really. As far as I'm aware, packagers usually roll their own setups and don't necessarily depend on the currently packaged rust version for a given OS. So depending on `stable` shouldn't be a problem :)

Feel free to bump the MSRV

**Review by Nukesor — CHANGES_REQUESTED** [review]
First up, thanks again :) 

Overall clean code structure, well documented and readable code.
The system setup was certainly more involved than I expected, but I guess it also gives the service itself a lot more control over itself.

A few remarks, mostly nitpicks and call for more documentation

Disclaimer:
I really didn't read up on the internals on how Windows services and their privileges work. I only read through the code, which I obviously didn't always fully understand because of that. So I'll just trust you and hope that you'll be available if anything breaks in this part :grin:

File: pueue/src/bin/pueued.rs [hunk 2]
```diff
@@ -78,7 +125,20 @@ fn fork_daemon(opt: &CliArguments) -> Result<()> {
...
+    #[cfg(target_os = "windows")]
+    {
+        use std::os::windows::process::CommandExt;
+        const CREATE_NO_WINDOW: u32 = 0x08000000;
+        // create_no_window causes all children to not show a visible console window
+        // but it also apparently has the effect of DETACH_PROCESS in that it's no longer tied to
+        // the parent process
+        // https://learn.microsoft.com/en-us/windows/win32/procthread/process-creation-flags#flags
+        command.creation_flags(CREATE_NO_WINDOW);
+    }
```
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/bin/pueued.rs:139]: Yeah! The Windows docs actually aren't even that clear on this for some reason. They didn't even mention it affects all children, and that it also causes some kind of detaching. Found that out the hard way after so much testing :( (also thanks stack overflow)

File: pueue/src/daemon/service.rs [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+
+                // Logoff
+                ServiceControl::SessionChange(SessionChangeParam {
+                    reason: SessionChangeReason::SessionLogoff,
+                    ..
+                }) => {
+                    debug!("event logoff");
+                    spawner.stop();
+
+                    ServiceControlHandlerResult::NoError
```
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs:234]: So basically, what would happen here is that when a user logs off, Windows itself kills all user processes. The service has no effect on this to be honest, and it's just following what Windows already does.

File: pueue/src/daemon/service.rs [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+
+pub fn stop_service() -> Result<()> {
+    let manager_access = ServiceManagerAccess::CONNECT;
+    let service_manager = ServiceManager::local_computer(None::<&str>, manager_access)?;
+
+    let service_access = ServiceAccess::QUERY_STATUS | ServiceAccess::STOP;
+    let service = service_manager.open_service(SERVICE_NAME, service_access)?;
+
+    match service.query_status()?.current_state {
+        ServiceState::Stopped => println!("Successfully is already stopped"),
```
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs:164]: Ah! Looks like a typo got past me! This is supposed to say `Service is already stopped`

File: pueue/src/daemon/service.rs [hunk 5]
```diff
@@ -0,0 +1,639 @@
...
+                let current_exe = env::current_exe()?.to_string_lossy().to_string();
+
+                let mut command = format!(r#""{current_exe}" {arguments}"#)
+                    .encode_utf16()
+                    .chain(iter::once(0))
+                    .collect::<Vec<_>>();
+
+                // lpcommandline may modify the the cmd vec. max valid size is 1024, so
+                // I think it best to ensure 1024 bytes of capacity total exist just in case
+                command.reserve(1024 - (command.len() / 2));
```
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs:572]: `lpcommandline` is the third argument of [`CreateProcessAsUserW`](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessasuserw). From the docs it was not clear to me exactly _how much_ it was going to alter the string, so I put this here out of an abundance of caution to ensure no UB happens.

File: pueue/src/bin/pueued.rs [hunk 6]
```diff
@@ -5,15 +5,24 @@ use clap::Parser;
...
     let opt = CliArguments::parse();
 
     if opt.daemonize {
+        // Ordinarily this would be handled in clap, but they don't support conflicting args
+        // with subcommands
+        #[cfg(target_os = "windows")]
+        if opt.service.is_some() {
+            use clap::CommandFactory;
+            let mut cmd = CliArguments::command();
+            cmd.print_help()?;
```
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/bin/pueued.rs:22]: Hmm, what message would you prefer here? I can only think of `failed to print help` as that's what's relevant here.

File: pueue/src/daemon/service.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonsan) [reply pueue/src/daemon/service.rs:234]: @MolotovCherry Great explanation, thank you for that!

File: pueue/src/daemon/service.rs [hunk 7]
```diff
@@ -0,0 +1,639 @@
+use std::{
```
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs:1]: > Explain a few simple things, such as how windows services work (they need to be registered and are afterwards permanently known to the system) and how are they started/stopped.

Services are installed by an administrator user. There is an official Windows service manager one can use to start/stop/restart the service, though we also included this functionality in the binary (minus the restart one, but this can be emulated by the user anyways with 2 commands). On install, it is set by default to autostart, and the system starts the service every time you start the system (unless the user changed the setting to manual or disabled in the service manager).

When a service is installed, Windows calls a preset cli string (our binary + args) which communicates with the service api (hence why we have a special flag for this, to trigger the api communication). We use an absolute path, so, the binary cannot be moved. This is to avoid the need to add it to PATH (which is an annoyance for the common user).

When the system is started, it listens for events as you've said, and when a user logs in, spawns a pueued daemon as the user, and on logoff it kills it (this technically is a noop. Windows itself kills all user processes on logoff anyways. More will be explained in the module docs)

> From what I understand, the service is registered with the args to call the service run subcommand whenever the service is to be started, which then listens to certain system level events, such as shutdown/logout/logon.
> 
> On start, that pueued instance then goes ahead and spawns another pueued instance, which is then the actual pueued instance.

Basically yes! (Adding module docs soon)

[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs:31]: Module docs added, let me know if it's sufficient for you

File: pueue/src/bin/pueued.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/bin/pueued.rs]: Printed a specific message rather than the help. Let me know if this is good enough. :)

File: pueue/src/daemon/service.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs:264]: Explained this in the module docs.

File: pueue/src/daemon/service.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs:31]: Module docs added. Let me know if it's good enough for you!

File: pueue/src/daemon/service.rs [hunk 8]
```diff
@@ -0,0 +1,639 @@
...
+
+impl Drop for OwnedHandle {
+    fn drop(&mut self) {
+        if !self.0.is_invalid() {
+            _ = unsafe { CloseHandle(self.0) };
+        }
+    }
+}
+
+/// a child process. tries to kill the process when dropped
```
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs]: Done!

File: pueue/src/daemon/service.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs]: According to the docs, 
> The system adds a null character to the command line string to separate the file name from the arguments. This divides the original string into two strings for internal processing. 

So 1 extra character may be enough. So I reduced this to 10 bytes instead (should be enough). Once I find clearer explanation than this (these docs are still a bit confusing) it could be fixed with 1 or whatever.

**Comment by MolotovCherry** [comment]
> I really didn't read up on the internals on how Windows services and their privileges work. I only read through the code, which I obviously didn't always fully understand because of that. So I'll just trust you and hope that you'll be available if anything breaks in this part

To keep it simple, some api calls require certain privileges (which the windows docs state which one). As this is running as SYSTEM, the privileges for this are probably already granted by default. This is just me being explicit for safety's sake. I don't anticipate there's any way to break this actually.

File: pueue/src/daemon/service.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MolotovCherry) [reply pueue/src/daemon/service.rs]: Ok, got this resolved. It modifies the string _in-place_, so no buffer is needed. We can safely completely remove this portion of code.

**Review by Nukesor — APPROVED** [review]
Boi, that's what I call thorough documentation :D 

Thanks a lot, whoever will take a look at this in the future will have all the resources they need to understand what's going on :) Much appreciated!

I've no further remarks, lets merge this!
